libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/idp/SPRequestHandler.java

```diff
@@ -26,10 +26,131 @@
+        spConfig = buildSPConfig();
+    }
+
+    /**
+     * Launches the IDP app.
+     */
+    public void launchIDPApp() {
+        /*
+         * TODO: Launch the subclass of user switcher screen for IDP that selects user.
+         * This user should take the SPConfig extra coming from here and add the selected
+         * UserAccount extra to it and then launch IDCodeGeneratorActivity.
+         */
+    }
+
+    /**
+     * Handles the response from the IDP app.
+     *
+     * @param resultCode Result code.
+     * @param data Data returned from the IDP app.
+     */
+    public void handleIDPResponse(int resultCode, Intent data) {
+        if (resultCode == Activity.RESULT_CANCELED) {
+            final String error = data.getStringExtra(IDPCodeGeneratorActivity.ERROR_KEY);
+            handleError(error);
+        } else if (resultCode == Activity.RESULT_OK) {
+            final String code = data.getStringExtra(IDPCodeGeneratorActivity.CODE_KEY);
+            final String loginUrl = data.getStringExtra(IDPCodeGeneratorActivity.LOGIN_URL_KEY);
+            handleSuccess(code, loginUrl);
+        }
+    }
+
+    /**
+     * Returns the SP config associated with this app.
+     *
+     * @return SPConfig instance.
+     */
+    public SPConfig getSpConfig() {
+        return spConfig;
+    }
+
+    private SPConfig buildSPConfig() {
+        final BootConfig bootConfig = BootConfig.getBootConfig(SalesforceSDKManager.getInstance().getAppContext());
+        return new SPConfig(bootConfig.getRemoteAccessConsumerKey(), bootConfig.getOauthRedirectURI(),
+                codeChallenge, bootConfig.getOauthScopes(), loginUrl, null);
+    }
+
+    private void handleError(String error) {
+        SalesforceSDKLogger.e(TAG, "Error received from IDP app: " + error);
+    }
+
+    private void handleSuccess(String code, String loginUrl) {
+        new TokenEndpointTask(code, loginUrl).execute();
+    }
+
+    private class TokenEndpointTask extends AsyncTask<Void, Void, OAuth2.TokenEndpointResponse> {
+
+        private String code;
+        private String loginUrl;
+
+        public TokenEndpointTask(String code, String loginUrl) {
+            this.code = code;
+            this.loginUrl = loginUrl;
+        }
+
+        @Override
+        protected OAuth2.TokenEndpointResponse doInBackground(Void... nothings) {
+            OAuth2.TokenEndpointResponse tokenResponse = null;
+            try {
+                tokenResponse = OAuth2.getSPCredentials(HttpAccess.DEFAULT,
+                        URI.create(loginUrl), spConfig.getOauthClientId(), code, codeVerifier,
+                        spConfig.getOauthCallbackUrl());
+            } catch (Exception e) {
+                SalesforceSDKLogger.e(TAG, "Exception occurred while making token request", e);
+            }
+            return tokenResponse;
+        }
+
+        @Override
+        protected void onPostExecute(OAuth2.TokenEndpointResponse tokenResponse) {
+            /*
```

Comment:
Getting token response here. Will wire this to adding a user account in the next PR.